Applications/SpaceAnalyzer/TreeMapWidget.h

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <LibGUI/Frame.h>
+#include <LibGfx/Rect.h>
+
+namespace TreeMap {
+
+struct ITreeMapNode {
+    virtual String name() = 0;
+    virtual int64_t area() = 0;
+    virtual size_t num_children() = 0;
+    virtual ITreeMapNode& child_at(size_t i) = 0;
+    virtual void sort_children_by_area() = 0;
+};
+
+struct ITreeMap : public RefCounted<ITreeMap> {
+    virtual ~ITreeMap() { }
+    virtual ITreeMapNode* get_root() = 0;
+};
+
+class TreeMapWidget final : public GUI::Frame {
+    C_OBJECT(TreeMapWidget)
+public:
+    virtual ~TreeMapWidget() override;
+    Function<void()> on_path_change;
+    size_t path_size();
+    ITreeMapNode* get_nth_path_node(size_t n);
+    size_t get_viewpoint();
+    void set_viewpoint(size_t);
+    void set_tree(RefPtr<ITreeMap> tree);
+
+private:
+    TreeMapWidget();
+    virtual void paint_event(GUI::PaintEvent&) override;
+    virtual void mousedown_event(GUI::MouseEvent&) override;
+    virtual void doubleclick_event(GUI::MouseEvent&) override;
+    virtual void mousewheel_event(GUI::MouseEvent&) override;
+
+    bool area_can_contain_children(Gfx::IntRect ext);
+    bool area_can_contain_label(Gfx::IntRect ext);
+    void paint_cell(GUI::Painter*, ITreeMapNode&, Gfx::IntRect ext, int depth, Gfx::IntPoint query_pos, Vector<int>& out_path);
+    void paint_node(GUI::Painter*, ITreeMapNode&, Gfx::IntRect ext, int depth, Gfx::IntPoint query_pos, Vector<int>& out_path);
+    Gfx::IntRect paint_cell_frame(GUI::Painter*, ITreeMapNode&, Gfx::IntRect ext, int depth);
```

Comment:
Please pass `GUI::Painter` by reference, since `nullptr` is not a valid value for it.
Also, please pass `Gfx::IntPoint` by const reference.
Logically speaking, it feels weird for these to take the treemap node by mutable reference. I guess it's because you may need to sort it. Since the sorting is essentially a lazy operation, it might make sense to make sorting const-qualified and pass `const ITreeMapNode&` to these.

Reply:
Not sure what to do about this, the reason I use GUI::Painter *, is that these functions also get called by event handlers for mousedown and doubleclick in which case there is no painter.

I also either don't understand the last part of your comment or simply don't know how to do it. (Making sorting const-qualified that is.)

Reply:
> Not sure what to do about this, the reason I use GUI::Painter *, is that these functions also get called by event handlers for mousedown and doubleclick in which case there is no painter.

Aha! Well that's extremely confusing :) Something called "paint_something" should not be called by a mouse event handler, since there's no painting happening.

Can we split those functions up into helpers that both paint and mouse events can call, without having to pass a painter in?

> I also either don't understand the last part of your comment or simply don't know how to do it. (Making sorting const-qualified that is.)

You are currently passing in the node as a `TreeMapNode&`. This looks very suspicious, as painting, logically speaking, is a read-only operation. So I would expect it to take `const TreeMapNode&`.

Reply:
The layout algorithm is now split out into its own function, it is called by paint and mouse event handlers.
The lay_out_children function still takes a non const reference to TreeMapNode, because it needs to call sort_children_by_area. 

You mentioned making sorting const-qualified, I don't know how to do that.

Options I can think of:
- Do all sorting during tree construction, that would mean sorting nodes that are potentially never displayed (i.e. wasted work).
- Do sorting during layout into temporary storage, means no performance gains on subsequent layout calls.
- Keep as is.

Reply:
You can make sorting const-qualified and then `const_cast<TreeMapWidget*>(this)->...` for mutable access inside `sort_children_by_area()` :)